### ISSUE TITLE
Double backward support for v3

### ISSUE THREAD

**Issue by beam2d** [chainer/chainer#3147]
This is a todo list of supporting double backward in typical convnets and RNNs. The aim of this issue is to manage the tasks so that no work will duplicate.

The original list is from my colleague that uses double backward in his research (currently using another framework or manual derivation of the first grad). Please add a comment if you want to add another function to the list. Let me know if you want to work on the task (I'll write your name to the list so that it is obvious who is working on what).

- [x] `+`, `*` (#2970)
- [x] `linear` (#3099, @beam2d)
- [x] `convolution_2d` and `deconvolution_2d` (#3163, @beam2d)
- [x] `max_pooling_2d` (#3257, @mitmul)
- [x] `average_pooling_2d` (#3257, @mitmul)
- [x] `upsampling_2d` (#3257, @mitmul)
- [x] `unpooling_2d` (#3257, @mitmul)
- [x] `relu` (#3175, @delta2323)
- [x] `lrelu` (#3177, @niboshi)
- [x] `exp` ( #3199 @aonotas )
- [x] `tanh` (#3200, @unnonouno )
- [x] `softmax` (#3213, @takagi )
- [x] `sigmoid` (#3119, @keisuke-umezawa)
- [x] `lstm` (#3206, @unnonouno)
- [x] `embed_id` (#3183, @unnonouno)
- [x] `mean_squared_error` (#3194, @delta2323)
- [x] `softmax_cross_entropy` (#3296, @delta2323)
- [x] `batch_normalization` (#3275, @beam2d)
- [x] `layer_normalization` (#3219, @niboshi)
- [x] `flatten` (#3190, @okuta)
- [x] `concat` and `split_axis` (#3189, @okuta)
- [x] `reshape`, `expand_dims`, `broadcast_to`, and `sum` (#3188, @okuta)
- [x] `get_item` (#3243, @delta2323)
- [x] `log_softmax` (#3217, @delta2323)
- [x] `rollaxis` (#3306, @delta2323)
- [x] `select_item` (#3308, @delta2323)
- [x] `dropout` (#3356, @bonprosoft)

After we decide on the treatment of existing, optimized backward #3146, I will write a guideline to write new-style functions in v3, and then dispatch tasks to core dev members.

**Comment by beam2d** [comment]
Note that the list is not a promise; we may remove some functions from the list in case of shortage of time.

**Comment by beam2d** [comment]
I wrote a guideline to support double backward of existing functions: https://gist.github.com/beam2d/dd87d97b8d7581f95cd19b1de86c9f4d. See also #3146 for the discussions on the way to keep the current, optimized backward implementations

**Comment by beam2d** [comment]
Note: we have decided to postpone Chainer v3 RC until we add double backprop support for essential functions listed in this issue. Sorry for the inconvenience.

**Comment by beam2d** [comment]
All the listed functions now support double backprop. Thank you for all involved in the development!
